### PR TITLE
AArch64 fastpath proofs

### PR DESCRIPTION
**Pull request by Xaphiosis**
Before I do more cleanup, I think this could use some feedback. Please ignore the horrid commit history (for now), but advice welcome on whether I should squash the actual fastpath commits into 3 parts (defs, rewrite, C refinement), or only one commit.

In any case the proofs work and look better than the ARM/ARM_HYP ones at least.

### CODE REVIEW

**Comment by Xaphiosis** [comment]
rebased onto master, had a de-sync between seL4 versions

**Comment by lsf37** [comment]
Nice work. Personally I'd slightly prefer 3 commits, or if you want to minimise, maybe 2 (one for defs + equiv, and one for the C part), but I don't have strong feelings about it.

File: lib/MonadicRewrite.thy
Comment (lsf37) [inline lib/MonadicRewrite.thy:532]: Would it add any convenience later in the proofs if we provided two versions where `F` is instantiated (and the guard therefore simplifies a bit), or does this disappear anyway before the user really sees it?

File: proof/crefine/AARCH64/Fastpath_Defs.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_Defs.thy:45]:
```suggestion
    asid \<leftarrow> returnOk $ fst $ the $ capPTMappedAddress $ capCap $ cteCap newVTable;
```
(or leaving out the first `$` instead should also work)

File: proof/crefine/AARCH64/Fastpath_Defs.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_Defs.thy:134]:
```suggestion
    asid \<leftarrow> returnOk $ fst $ the $ capPTMappedAddress $ capCap $ cteCap newVTable;
```

File: proof/crefine/AARCH64/Fastpath_Equiv.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_Equiv.thy:1762]: Intentional blank line?

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:86]: I had completely forgotten that this was in here. Should indeed consolidate these at some point, but fine to leave here for now, esp since the FIXME makes it findable.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:239]: Is there any chance of making the fast path proofs generic over 32/64 bit or should we in this case just not worry about those specific kind of magic numbers? (specifically 32 and 64)

Comment (lsf37) [reply proof/crefine/AARCH64/Fastpath_C.thy:239]: Looks like writing things out makes them clearer: yes, I think it is worth getting rid of the magic numbers, because it might at least make some parts of the fast path proofs textually equal or more similar, which makes it easier to apply simple-minded patches like renaming etc.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:328]:
```suggestion
          apply (rule ccorres_move_array_assertion_cnode_ctes ccorres_move_c_guard_cte
                 | csymbr)+
```

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:406]: What's our style policy for the `in`? Should this be the following?
```suggestion
          apply (rule_tac P'="{s. cap_' s = getCTE_cap}"
                          in ccorres_from_vcg_throws[where P=\<top>])
```

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:1058]: maybe add a FIXME on this one for removing the magic numbers

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:1823]: Looks perfectly legit to me, i.e. I don't think there's a significantly better way to do this.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [inline proof/crefine/AARCH64/Fastpath_C.thy:2501]: We could try to remove the simp rule that turns None into Ex Some. With that, clarsimp might be fine.

**Review by lsf37 — APPROVED** [review]
Very cool. Modulo the small stuff above and commit squashing, I think this is ready to merge. It's still gigantic, but it looks to me like proof quality has improved a bit compared to the other fast path versions.

We should add a commit that removes the `CREFINE_QUICK_AND_DIRTY` from the Makefile.

File: lib/MonadicRewrite.thy
Comment (Xaphiosis) [reply lib/MonadicRewrite.thy:532]: I thought about it, but didn't see any point. If you apply this rule, you end up with a wp goal later on, which will be cleared with a `simp` given any specific value of `F`. If an extreme situation, they can do `where F=..., simplified`, but again, not seeing it.

File: proof/crefine/AARCH64/Fastpath_Equiv.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_Equiv.thy:1762]: oops

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:239]: I think making them generic in 32/64 is not feasible, but obvious mentions of machine word lengths are something we should strive for, hence the FIXME. The issue is, if you want this merged ASAP, do I go doing these fixes now or after?

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:406]: I asked about this in mattermost before I did these, and people linked me to the discussion where we had 3 acceptable versions I think (rhs `and`: the `in` is like you point out; lhs `and` with `in` indented same as the `and`; the `and/in` right-aligned with the `c` in `rule_tac` (Michael style?)). I'm a bit confused at this point. I was confused before as well, because I started the original discussion.
Worse yet, trying to align the `in` so that the `n` is under the `c`, when seen in a normal text editor (or diff), results in the `n` being under the space instead.
Advice welcome.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:1058]: I put a comment right above which says this is copied by whatever shows up in `fastpath_mi_check_spec`, which I think are bitfield gen numbers? I have no clue what they are.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:2501]: There's still the issue of pairs, which would need to be folded back into `fst` and `snd`. Do we have tech to handle that?
There's also what Corey wrote on the wp schematic issue, where we might have a wrong rule in the `wp_comb` set.

File: proof/crefine/AARCH64/Fastpath_Defs.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_Defs.thy]: done

File: proof/crefine/AARCH64/Fastpath_Defs.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_Defs.thy]: done

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy]: done

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy]: removed the FIXME

**Comment by Xaphiosis** [comment]
> Very cool. Modulo the small stuff above and commit squashing, I think this is ready to merge. It's still gigantic, but it looks to me like proof quality has improved a bit compared to the other fast path versions.

I'd like your reply to my replies (maybe I catch you tonight) before resolving the rest. I also have some questions:
* do you want to merge this before `AARCH64 FIXME`s are resolved, and then we sort them out later / after the deadline? You seem to have a plan w.r.t. the timing, but I think I didn't pick up on it
* bike shed: `asid_has_vmid` name: if at some point we do RISCV64 fastpath, this name makes no sense there, and it's more than the VMID anyway. Should I do the rename as in the FIXME? A better name? `asid_has_vspace`? Leave it for now?
 
> We should add a commit that removes the `CREFINE_QUICK_AND_DIRTY` from the Makefile.

Good point, added that to the commits.

I'm not sure whether anyone else is going to look at this, but I will wait with squashing until tonight when we can hopefully resolve/discuss the remaining issues.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (corlewis) [reply proof/crefine/AARCH64/Fastpath_C.thy:2500]: To clarify, if I'm right in what I thought was going on with the `wp_comb` rule then it would just help with the problem you ran into of a schematic assumption being produced and then unified with `False`. The actual proof would still require either your careful rule application or the careful adjustment of the simp set like you two are talking about.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (michaelmcinerney) [inline proof/crefine/AARCH64/Fastpath_C.thy:369]: Intentional newline?

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (corlewis) [reply proof/crefine/AARCH64/Fastpath_C.thy:406]: Hmm, I guess our original discussion only considered the case with multiple instantiations. Sorry to add yet another version, but I thought our preference from that discussion was what I think you're describing with the right-aligned Michael style. That would make it
```suggestion
          apply (rule_tac P'="{s. cap_' s = getCTE_cap}"
                       in ccorres_from_vcg_throws[where P=⊤])
```

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (michaelmcinerney) [inline proof/crefine/AARCH64/Fastpath_C.thy:719]: Is this the `cteSize` or something that you were asking about on Mattermost? Is there any benefit of using that instead of the bare number here?

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (michaelmcinerney) [inline proof/crefine/AARCH64/Fastpath_C.thy:1905]: Another newline here, maybe unintentional

**Review by michaelmcinerney — APPROVED** [review]
Nice. Just a few nitpicks from me

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Fastpath_C.thy:239]: As long as the tag is there it's fine to do later.

Sorry for the confusing comment: initially I was thinking we could just remove the tag, because there is no point anyway, but then I discovered what the point is. So everything is fine and no action required for the PR.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Fastpath_C.thy:406]: Ok, I'm glad I'm not the only one who is confused :-)

I propose we leave this one as is and figure out separately what we really want for this case.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Fastpath_C.thy:2500]: Pairs being split should be handled correctly by `wpfix`, so I think removing the None/Some rule has a chance in this case (it won't fix the general problem). It has a tag already, so we're fine to merge, I was just curious what people's thoughts were on this one. If I get to it in the cleanup run I might try out removing the rule and see what happens.

**Comment by lsf37** [comment]
> > Very cool. Modulo the small stuff above and commit squashing, I think this is ready to merge. It's still gigantic, but it looks to me like proof quality has improved a bit compared to the other fast path versions.
> 
> I'd like your reply to my replies (maybe I catch you tonight) before resolving the rest. I also have some questions:
> 
> * do you want to merge this before `AARCH64 FIXME`s are resolved, and then we sort them out later / after the deadline? You seem to have a plan w.r.t. the timing, but I think I didn't pick up on it

I think we should merge this one first, so the proof is properly finished and `quick_and_dirty` is switched off. If we're lucky, we can rebase the existing FIXME cleanup PRs over it and they still work -- then I'd merge these too. If they need fixing, we can just see how far we get, they are less critical.

> * bike shed: `asid_has_vmid` name: if at some point we do RISCV64 fastpath, this name makes no sense there, and it's more than the VMID anyway. Should I do the rename as in the FIXME? A better name? `asid_has_vspace`? Leave it for now?

It's ok to leave it for now. The proposed `asid_has_vspace` is not bad, so I'd take that one when we get to it. The predicate will likely need changing for RISCV, since it does ASIDs differently, but the name `asid_has_vspace` would still fit. I guess it's "vspace with potentially additional data", but that is a bit long.

> > We should add a commit that removes the `CREFINE_QUICK_AND_DIRTY` from the Makefile.
> 
> Good point, added that to the commits.

👍
 
> I'm not sure whether anyone else is going to look at this, but I will wait with squashing until tonight when we can hopefully resolve/discuss the remaining issues.

We're good to merge from my side after squashing.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:406]: This looks like I'm carrying some kind of rendering issue with even placing the `in` in the right place, because the *intended* placement was where @corlewis said it should be. I'll fix those misalignments with vim for now so it's like what Corey said, and then I'll create an issue so we sort this out and can put it in the style guide.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:2500]: Ok, that would be nice, but I think the wp_comb thing also should happen if we're doing cleanup at some point, because the combo of wp, simp, vcg is a disaster to work with otherwise. When you say "when I get to it in the cleanup run", does that mean you've assumed responsibility for the doing entire cleanup run?

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:1905]: yeah, better without it

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy:719]: Yeah, this is the part that refers to maybe fixing it:
```
(* FIXME AARCH64 in this file, we see ptr + 0x20 * tcbVTableSlot a lot, where the 0x20 is
   2 ^ cte_level_bits
   This used to be 0x10 on 32-bit, so unlikely to change soon, but still might be worth a cleanup
   or abbreviation. This is one way it could work (and can go outside Arch locale): *)
declare cte_level_bits_def[code]
value_abbreviation (input) cte_size "(2::machine_word) ^ cte_level_bits"
```

It's not really clear if an input abbreviation only will be helpful here. Might be better to go with a constant, but it was hard to wrangle it in the proofs. Still not sure about this one, but yeah it's what I was asking about in mattermost

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (Xaphiosis) [reply proof/crefine/AARCH64/Fastpath_C.thy]: OK it seems in the final ccorres proofs I wasn't making the misalignment mistakes, so maybe they didn't come from me. It's still not entirely consistent w.r.t. the `in` placement so TODO: make issue.
Fix pushed.

**Comment by Xaphiosis** [comment]
Squashed and pushed. Might be worth waiting for the test results on this one as it adds the quick_and_dirty removal from the Makefile.

File: proof/crefine/AARCH64/Fastpath_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Fastpath_C.thy:2498]: I was planning to volunteer to do a first pass, but I wouldn't say I have assumed responsibility for all of the cleanup :-) There will probably be a whole bunch of things that need some insight and where we might want to discuss more, it'd be nice to do these together in some way.

**Comment by lsf37** [comment]
Happy with the squash and commit messages. Will wait for the tests to pass before merging.

**Pull request merged**
